src/test/kotlin/com/zepben/auth/client/ZepbenTokenFetcherTest.kt

```diff
@@ -0,0 +1,252 @@
+
+package com.zepben.auth.client
+
+import com.zepben.auth.common.AuthException
+import com.zepben.auth.common.AuthMethod
+import com.zepben.auth.common.StatusCode
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.AfterEach
+
+import org.junit.jupiter.api.Assertions.*
+import org.junit.jupiter.api.Test
+import com.zepben.testutils.vertx.TestHttpServer
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.core.IsEqual.equalTo
+import org.mockito.Mockito.*
+import com.zepben.testutils.auth.TOKEN
+
+import java.net.http.HttpClient
+import java.net.http.HttpResponse
+
+
+internal class ZepbenTokenFetcherTest {
+    private var server: TestHttpServer? = null
+    private var port = 8080
+
+    private val client: HttpClient = mock(HttpClient::class.java)
+    private val response = mock(HttpResponse::class.java)
+
+    @BeforeEach
+    fun beforeEach() {
+        server = TestHttpServer()
+        port = server!!.listen()
+        doReturn(response).`when`(client).send(any(), any(HttpResponse.BodyHandler::class.java))
+    }
+
+    @AfterEach
+    fun afterEach() {
+        server?.close()
+    }
+
+    @Test
+    fun testCreateTokenFetcherSuccess() {
+        doReturn(StatusCode.OK.code).`when`(response).statusCode()
+        doReturn(
+            "{\"authType\": \"AUTH0\", \"audience\": \"test_audience\", \"issuer\": \"test_issuer\"}"
+        ).`when`(response).body()
+
+        val tokenFetcher = createTokenFetcher("https://testaddress", confClient = client, authClient = client)
+        verify(client, times(1)).send(any(), any(HttpResponse.BodyHandler::class.java))
+        assertThat(tokenFetcher?.audience, equalTo("test_audience"))
+        assertThat(tokenFetcher?.issuerDomain, equalTo("test_issuer"))
+    }
+
+    @Test
+    fun testCreateTokenFetcherNoAuth() {
+        doReturn(StatusCode.OK.code).`when`(response).statusCode()
+        doReturn(
+            "{\"authType\": \"NONE\", \"audience\": \"\", \"issuer\": \"\"}"
+        ).`when`(response).body()
+
+        val tokenFetcher = createTokenFetcher("https://testaddress", confClient = client, authClient = client)
+        verify(client, times(1)).send(any(), any(HttpResponse.BodyHandler::class.java))
+        assertThat(tokenFetcher, equalTo(null))
+    }
+
+    @Test
+    fun testCreateTokenFetcherBadResponse() {
+        doReturn(StatusCode.NOT_FOUND.code).`when`(response).statusCode()
+        doReturn("Not found").`when`(response).body()
+
+        val exception = assertThrows(AuthException::class.java) {
+            createTokenFetcher("https://testaddress", confClient = client, authClient = client)
+        }
+        verify(client, times(1)).send(any(), any(HttpResponse.BodyHandler::class.java))
+        assertThat(exception.statusCode, equalTo(StatusCode.NOT_FOUND.code))
+        assertThat(exception.message, equalTo("https://testaddress responded with error: 404 - Not found"))
```

Comment:
Can use the zepben.exception checker here (I have not run the following so might not be quite correct):
```
        expect {
            createTokenFetcher("https://testaddress", confClient = client, authClient = client)
        }.toThrow(AuthException::class.java)
            .withMessage("https://testaddress responded with error: 404 - Not found")
            .exception
            .apply {
                verify(client, times(1)).send(any(), any(HttpResponse.BodyHandler::class.java))
                assertThat(statusCode, equalTo(StatusCode.NOT_FOUND.code))
            }
```

@kgreav Another task that I have been meaning to do that you can get @Lax125 to work on is making the above work nicer with Kotlin. e.g. `.toThrow<AuthException>()` rather than `.toThrow(AuthException::class.java)`